core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java

```diff
@@ -231,7 +236,21 @@ public void initializeConfiguration(
         builder.withWrapper(ExpandingConfigSource::new);
         builder.addDefaultSources();
         final ApplicationPropertiesConfigSource.InJar inJar = new ApplicationPropertiesConfigSource.InJar();
-        builder.withSources(inJar);
+        if (applicationInfo.isPresent()) {
+            final HashMap<String, String> appStateProperties = new HashMap<>();
+            final ApplicationInfo defaultValues = applicationInfo.get().getApplicationInfo();
+            // these are the default values that will be used for ApplicationConfig
+            appStateProperties.put("quarkus.application.name", defaultValues.getName());
+            appStateProperties.put("quarkus.application.version", defaultValues.getVersion());
+
+            builder.withSources(
+                    inJar,
+                    // the ordinal of 240 because it's slightly lower priority then what inJar uses
+                    new PropertiesConfigSource(appStateProperties, "appState", 240));
```

Comment:
This isn't the right way to set default values.

The best way to do it is to just use the `defaultValue` property on the `@ConfigItem`.  The second-best way is to just implement whatever default behavior you want when the properties are not present in the config object.  The third-best way would be to produce default value build items.  Adding a config source doesn't even make the list unfortunately. ;-)

Reply:
If you (or anyone) find yourself modifying the core configuration code or the core mojos or main methods in order to add a configuration switch, then a wrong turn was taken at some point.  The configuration and build item infrastructures exist so that we don't have to do these things.

Reply:
I really wanted to use `defaultValue`, but the problem is that default needs to be dynamic, in the sense that it depends on what is in the build.
My idea was that the "client" processors would just inject the `ApplicationConfig` and not have to worry about how to set default values if users don't.

So just to verify, what your proposing is that the processors would inject `ApplicationConfig` which would contain whatever the users set and then would also have to inject the default values in some other way?

Reply:
The build step (singular) would accept the config and produce the build item with the actual app name in it.  The build step would also calculate the default value for the build item in the event that it was not given in the config.  Everything which needs the application name and/or version would consume the build item.

Reply:
OK, got it thanks for the explanation! 
So I implemented it the other way around. But I still don't understand how I would pull the default values unless I am pushing them in when setting up the Augmentor.

Reply:
You'd have to push them in somehow.  One somewhat hacky option would be to make the config `defaultValue` be equal to e.g. `${artifactId:}`, and then make sure that property is set when calling the augmentor.

TBH we might need a better general mechanism for default config that comes from the calling environment, rather than having one-offs.  That could also encapsulate things like the classes root, wiring dir, etc., allowing those things to be driven by config as well.

Reply:
The kubernetes extension would definitely need those values as well (by the means of injecting a build item that contains the final values).
I really like the idea of having these values settable in `application.properties`, but I think it would require a fair amount of work. WDYT?

Reply:
As long as the Kube extension just consumed that build item, and assuming it was produced as I described above, it should just work with a minimum of effort.

Reply:
Having the extension read the config should be easy once everything is in place. The difficulty I see is having all the config (like wiring classes) driven by `application.properties`

Reply:
Ah I see.  Well it doesn't make sense to force 100% of everything through properties.  It's more like two polar endpoints: at one end, all application-wide config must be driven by `application.properties`, and at the other end, any per-class or per-member behavior should be driven by annotations.  In between is where we consider factors to determine if it makes sense to make some configuration be driven one way and some the other, or whether to allow both.

No matter how it goes from the input stage, the next stage is to feed these into build items, so that consumers don't actually care where they came from.

Reply:
👍 I will rework this PR starting tomorrow so we have something more concrete to comment on 😊